File: src/main/java/com/github/nsc/de/compiler/parser/node/expression/ExpressionNode.java
package com.github.nsc.de.compiler.parser.node.expression;

import com.github.nsc.de.compiler.parser.node.ValuedNode;

public abstract class ExpressionNode implements ValuedNode {
    private final ValuedNode left;
    private final ValuedNode right;

    public ExpressionNode(ValuedNode left, ValuedNode right) {
        this.left = left;
        this.right = right;
    }

    public abstract char getOperator();

    public ValuedNode getLeft() {
        return left;
    }
    public ValuedNode getRight() {
        return right;
    }

    @Override
    public String toString() {
        return "("+(getLeft() != null && !(getLeft() instanceof NumberNode && ((NumberNode)getLeft()).getNumber() == 0) ? getLeft() : "")+getOperator()+getRight()+")";
    }
}

File: shake/compiler/parser/src/commonMain/kotlin/io/github/shakelang/shake/parser/node/expression/ExpressionNode.kt
package io.github.shakelang.shake.parser.node.expression

import io.github.shakelang.shake.parser.node.ValuedNode
import io.github.shakelang.parseutils.characters.position.PositionMap

@Suppress("unused")
abstract class ExpressionNode(
    map: PositionMap,
    val left: ValuedNode,
    val right: ValuedNode,
    val operatorPosition: Int
) : ValuedNode(map) {
    abstract val operator: Char
}